docs/userguide/datafiles.rst

```diff
@@ -12,6 +12,12 @@ Setuptools focuses on this most common type of data files and offers three ways
 of specifying which files should be included in your packages, as described in
 the following sections.
 
+The user may want to remove the automatically generated PKG-INFO, \*.egg-info,
+build, and dist folders when modifying the configuration files or doing
+structural changes in the directory/file structure. Failure to do so may result
+in an sdist that contains datafiles which are not in a simultaneously built
+wheel, and which will not be installed even if directly installing the sdist.
```

Comment:
Maybe something along the lines of:

```suggestion
.. note::
   Setuptools automatically creates a few directories to host build artefacts
   and cache files, such as ``build``, ``dist``, ``*.egg-info``.
   While cache is useful to speed up incremental builds, in some edge cases it might become stale.
   If you feel that caching is causing problems to your build, specially after changes in
   configuration or in the directory/file structure., consider removing
   ``build``, ``dist``, ``*.egg-info`` [#PKG-INFO] before rebuilding or
   reinstalling your project.

.. [#PKG-INFO]
   When working from an existing sdist (e.g. for patching), you might also consider removing
   the `PKG-INFO` file to force its recreation.
``` 

I suggest putting the `PKG-INFO` part in a footnote because those are not day-to-day operations for a standard setuptools user. It is something very advanced.


Nevertheless, I think it would make more sense to add this content as a dedicated session in the document (with a heading title) more towards the end. The reason being that we should organise the information to have a nice reading flow. I think that the "attention scenarios" and workaround/debugging tips should come after the configuration methods are explained.

Reply:
Another thing that might make sense is to reference this content from the part of the document we mention `MANIFEST.in`.

Reply:
The way that you have written it makes sense. I will rework what I have to be more like that, and I agree that flow is important. I just have two issues here:

1. A user who thinks they have a simple situation and wants to get back to something else won't read the whole page, they will get far enough to have something that might work and stop there. I think it is important to quickly warn users who have already built a package, that they should look at the note at the bottom. If it is a quick note it won't interfere with those who are configuring a brand new package who won't be affected by stale artefacts.
2. I got to the point of having created and uploaded several of my own packages without really understanding the differences between an `sdist` and a `wheel` beyond naming and archive format. Specifically, I was convinced that if an editable install had the needed files that I could look in the `sdist` to see if the files were there as well. I hadn't seen any of the documentation that would have said that `sdist` archives are expected to have files that will not be installed, and will not produce warnings or errors if there are files that are in the `sdist` that are not used in building what does get installed. This page is one of the first ones that I ran into when I realized that I needed to get files that weren't Python source to be installed, and I suspect others at my level of understanding will end up here as well. I recognize that it does interrupt the flow to be talking about the difference between what gets packaged and what gets installed here, but there should at least be a link to somewhere that it is explained, preferably by someone who understands it better than I do.